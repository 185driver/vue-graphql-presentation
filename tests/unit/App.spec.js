import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueApollo from 'vue-apollo'
import Vuetify from 'vuetify'
import AppComponent from '@/App.vue'
import { createMockClient } from 'mock-apollo-client'
import allHeroesQuery from '@/graphql/allHeroes.query.gql'
import addHeroMutation from '@/graphql/addHero.mutation.gql'
import deleteHeroMutation from '@/graphql/deleteHero.mutation.gql'
import VueHero from '@/components/VueHero'

const heroListMock = {
  data: {
    allHeroes: [
      {
        github: 'test-github',
        id: '-1',
        image: 'image-link',
        name: 'Anonymous Vue Hero',
        twitter: 'some-twitter',
      },
      {
        github: 'test-github2',
        id: '-2',
        image: 'image-link2',
        name: 'another Vue Hero',
        twitter: 'some-twitter2',
      },
    ],
  },
}

const newHeroMock = {
  name: 'New Hero',
  github: '1000-contributions-a-day',
  twitter: 'new-hero',
  image: 'img.jpg',
}

const newHeroMockResponse = {
  data: {
    addHero: {
      __typename: 'Hero',
      id: '123',
      ...newHeroMock,
    },
  },
}

const localVue = createLocalVue()
localVue.use(VueApollo)
localVue.use(Vuetify)

describe('App component', () => {
  let wrapper
  let mockClient
  let apolloProvider
  let requestHandlers

  const createComponent = (handlers) => {
    requestHandlers = {
      allHeroesQueryHandler: jest.fn().mockResolvedValue(heroListMock),
      addHeroMutationHandler: jest.fn().mockResolvedValue(newHeroMockResponse),
      deleteHeroMutationHandler: jest
        .fn()
        .mockResolvedValue({ data: { deleteHero: true } }),
      ...handlers,
    }
    mockClient = createMockClient()
    mockClient.setRequestHandler(
      allHeroesQuery,
      requestHandlers.allHeroesQueryHandler
    )
    mockClient.setRequestHandler(
      addHeroMutation,
      requestHandlers.addHeroMutationHandler
    )
    mockClient.setRequestHandler(
      deleteHeroMutation,
      requestHandlers.deleteHeroMutationHandler
    )

    apolloProvider = new VueApollo({
      defaultClient: mockClient,
    })
    wrapper = shallowMount(AppComponent, {
      localVue,
      apolloProvider,
    })
  }

  afterEach(() => {
    wrapper.destroy()
    mockClient = null
    apolloProvider = null
  })

  it('renders a loading block when query is in progress', () => {
    createComponent()

    expect(wrapper.find('.test-loading').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a list of two heroes when query is resolved', async () => {
    createComponent()

    await wrapper.vm.$nextTick()

    expect(wrapper.find('.test-loading').exists()).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAllComponents(VueHero)).toHaveLength(2)
  })

  it('renders a message about no heroes when heroes list is empty', async () => {
    createComponent({
      allHeroesQueryHandler: jest
        .fn()
        .mockResolvedValue({ data: { allHeroes: [] } }),
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.find('.test-loading').exists()).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.test-empty-list').exists()).toBe(true)
  })

  it('renders error if query fails', async () => {
    createComponent({
      allHeroesQueryHandler: jest
        .fn()
        .mockRejectedValue(new Error('GraphQL error')),
    })

    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(wrapper.find('.test-loading').exists()).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.test-error').exists()).toBe(true)
  })

  it('adds a new hero to cache on addHero mutation', async () => {
    createComponent()
    wrapper.setData({
      ...newHeroMock,
    })

    await wrapper.vm.$nextTick()
    wrapper.vm.addHero()

    expect(requestHandlers.addHeroMutationHandler).toHaveBeenCalledWith({
      hero: {
        ...newHeroMock,
      },
    })

    await wrapper.vm.$nextTick()

    expect(
      mockClient.cache.readQuery({ query: allHeroesQuery }).allHeroes
    ).toHaveLength(3)

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAllComponents(VueHero)).toHaveLength(3)
  })

  it('deletes a hero from cache correctly', async () => {
    createComponent()

    await wrapper.vm.$nextTick()

    wrapper
      .findAllComponents(VueHero)
      .at(0)
      .vm.$emit('deleteHero', heroListMock.data.allHeroes[0].name)

    expect(requestHandlers.deleteHeroMutationHandler).toHaveBeenCalledWith({
      name: 'Anonymous Vue Hero',
    })

    await wrapper.vm.$nextTick()

    expect(
      mockClient.cache.readQuery({ query: allHeroesQuery }).allHeroes
    ).toHaveLength(1)
    expect(
      mockClient.cache
        .readQuery({ query: allHeroesQuery })
        .allHeroes.some((hero) => hero.name === 'Anonymous Vue Hero')
    ).toBe(false)

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAllComponents(VueHero)).toHaveLength(1)
  })
})
