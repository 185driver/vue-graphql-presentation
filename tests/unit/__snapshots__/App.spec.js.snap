// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component adds a new hero to cache on addHero mutation 1`] = `
<v-app-stub id="app">
  <v-toolbar-stub fixed="true" app="true" color="teal darken-2" dark="true" scrollthreshold="300"><img src="./assets/vue-graphql.png" alt="Logo" class="logo">
    <v-toolbar-title-stub class="hidden-xs-only">Vue Heroes</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-content-stub tag="main" class="teal lighten-3">
    <v-container-stub tag="div" fluid="" fill-height="" grid-list-md="" class="app-container white">
      <v-layout-stub tag="div" wrap="" class="hero-cards-layout test-heroes-list">
        <vue-hero-stub hero="[object Object]"></vue-hero-stub>
        <vue-hero-stub hero="[object Object]"></vue-hero-stub>
        <vue-hero-stub hero="[object Object]"></vue-hero-stub>
        <v-flex-stub tag="div" md3="" xs12="">
          <v-card-stub tag="div" height="100%" class="centered test-saving">
            <v-progress-circular-stub indeterminate="true" rotate="0" size="32" width="4" value="0"></v-progress-circular-stub>
          </v-card-stub>
        </v-flex-stub>
      </v-layout-stub>
      <!---->
      <v-dialog-stub contentclass="" maxwidth="none" origin="center center" width="800" transition="dialog-transition">
        <v-btn-stub color="teal" activeclass="v-btn--active" tag="button" dark="true" type="button">Add Hero</v-btn-stub>
        <v-card-stub tag="div" class="dialog-form">
          <v-form-stub>
            <v-text-field-stub rules="v => !!v || 'Name is required'" label="Name" required="" value=""></v-text-field-stub>
            <v-text-field-stub label="Image Link" value=""></v-text-field-stub>
            <v-text-field-stub label="Twitter" value=""></v-text-field-stub>
            <v-text-field-stub label="Github" value=""></v-text-field-stub>
          </v-form-stub>
          <v-card-actions-stub>
            <v-btn-stub activeclass="v-btn--active" disabled="true" tag="button" type="button" class="test-submit">submit</v-btn-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-dialog-stub>
    </v-container-stub>
  </v-content-stub>
  <v-footer-stub app="true" color="teal darken-2" height="32"><span class="footer-text text-xs-center white--text">Â© 2019</span></v-footer-stub>
</v-app-stub>
`;

exports[`App component deletes a hero from cache correctly 1`] = `
<v-app-stub id="app">
  <v-toolbar-stub fixed="true" app="true" color="teal darken-2" dark="true" scrollthreshold="300"><img src="./assets/vue-graphql.png" alt="Logo" class="logo">
    <v-toolbar-title-stub class="hidden-xs-only">Vue Heroes</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-content-stub tag="main" class="teal lighten-3">
    <v-container-stub tag="div" fluid="" fill-height="" grid-list-md="" class="app-container white">
      <v-layout-stub tag="div" wrap="" class="hero-cards-layout test-heroes-list">
        <vue-hero-stub hero="[object Object]"></vue-hero-stub>
        <!---->
      </v-layout-stub>
      <!---->
      <v-dialog-stub contentclass="" maxwidth="none" origin="center center" width="800" transition="dialog-transition">
        <v-btn-stub color="teal" activeclass="v-btn--active" tag="button" dark="true" type="button">Add Hero</v-btn-stub>
        <v-card-stub tag="div" class="dialog-form">
          <v-form-stub>
            <v-text-field-stub rules="v => !!v || 'Name is required'" label="Name" required="" value=""></v-text-field-stub>
            <v-text-field-stub label="Image Link" value=""></v-text-field-stub>
            <v-text-field-stub label="Twitter" value=""></v-text-field-stub>
            <v-text-field-stub label="Github" value=""></v-text-field-stub>
          </v-form-stub>
          <v-card-actions-stub>
            <v-btn-stub activeclass="v-btn--active" disabled="true" tag="button" type="button" class="test-submit">submit</v-btn-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-dialog-stub>
    </v-container-stub>
  </v-content-stub>
  <v-footer-stub app="true" color="teal darken-2" height="32"><span class="footer-text text-xs-center white--text">Â© 2019</span></v-footer-stub>
</v-app-stub>
`;

exports[`App component renders a list of two heroes when query is resolved 1`] = `
<v-app-stub id="app">
  <v-toolbar-stub fixed="true" app="true" color="teal darken-2" dark="true" scrollthreshold="300"><img src="./assets/vue-graphql.png" alt="Logo" class="logo">
    <v-toolbar-title-stub class="hidden-xs-only">Vue Heroes</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-content-stub tag="main" class="teal lighten-3">
    <v-container-stub tag="div" fluid="" fill-height="" grid-list-md="" class="app-container white">
      <v-layout-stub tag="div" wrap="" class="hero-cards-layout test-heroes-list">
        <vue-hero-stub hero="[object Object]"></vue-hero-stub>
        <vue-hero-stub hero="[object Object]"></vue-hero-stub>
        <!---->
      </v-layout-stub>
      <!---->
      <v-dialog-stub contentclass="" maxwidth="none" origin="center center" width="800" transition="dialog-transition">
        <v-btn-stub color="teal" activeclass="v-btn--active" tag="button" dark="true" type="button">Add Hero</v-btn-stub>
        <v-card-stub tag="div" class="dialog-form">
          <v-form-stub>
            <v-text-field-stub rules="v => !!v || 'Name is required'" label="Name" required="" value=""></v-text-field-stub>
            <v-text-field-stub label="Image Link" value=""></v-text-field-stub>
            <v-text-field-stub label="Twitter" value=""></v-text-field-stub>
            <v-text-field-stub label="Github" value=""></v-text-field-stub>
          </v-form-stub>
          <v-card-actions-stub>
            <v-btn-stub activeclass="v-btn--active" disabled="true" tag="button" type="button" class="test-submit">submit</v-btn-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-dialog-stub>
    </v-container-stub>
  </v-content-stub>
  <v-footer-stub app="true" color="teal darken-2" height="32"><span class="footer-text text-xs-center white--text">Â© 2019</span></v-footer-stub>
</v-app-stub>
`;

exports[`App component renders a loading block when query is in progress 1`] = `
<v-app-stub id="app">
  <v-toolbar-stub fixed="true" app="true" color="teal darken-2" dark="true" scrollthreshold="300"><img src="./assets/vue-graphql.png" alt="Logo" class="logo">
    <v-toolbar-title-stub class="hidden-xs-only">Vue Heroes</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-content-stub tag="main" class="teal lighten-3">
    <v-container-stub tag="div" fluid="" fill-height="" grid-list-md="" class="app-container white">
      <div class="test-loading">
        Loading...
      </div>
    </v-container-stub>
  </v-content-stub>
  <v-footer-stub app="true" color="teal darken-2" height="32"><span class="footer-text text-xs-center white--text">Â© 2019</span></v-footer-stub>
</v-app-stub>
`;

exports[`App component renders a message about no heroes when heroes list is empty 1`] = `
<v-app-stub id="app">
  <v-toolbar-stub fixed="true" app="true" color="teal darken-2" dark="true" scrollthreshold="300"><img src="./assets/vue-graphql.png" alt="Logo" class="logo">
    <v-toolbar-title-stub class="hidden-xs-only">Vue Heroes</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-content-stub tag="main" class="teal lighten-3">
    <v-container-stub tag="div" fluid="" fill-height="" grid-list-md="" class="app-container white">
      <v-layout-stub tag="div" wrap="" class="hero-cards-layout test-heroes-list">
        <!---->
      </v-layout-stub>
      <h3 class="headline test-empty-list">
        No heroes found ðŸ˜­
      </h3>
      <v-dialog-stub contentclass="" maxwidth="none" origin="center center" width="800" transition="dialog-transition">
        <v-btn-stub color="teal" activeclass="v-btn--active" tag="button" dark="true" type="button">Add Hero</v-btn-stub>
        <v-card-stub tag="div" class="dialog-form">
          <v-form-stub>
            <v-text-field-stub rules="v => !!v || 'Name is required'" label="Name" required="" value=""></v-text-field-stub>
            <v-text-field-stub label="Image Link" value=""></v-text-field-stub>
            <v-text-field-stub label="Twitter" value=""></v-text-field-stub>
            <v-text-field-stub label="Github" value=""></v-text-field-stub>
          </v-form-stub>
          <v-card-actions-stub>
            <v-btn-stub activeclass="v-btn--active" disabled="true" tag="button" type="button" class="test-submit">submit</v-btn-stub>
          </v-card-actions-stub>
        </v-card-stub>
      </v-dialog-stub>
    </v-container-stub>
  </v-content-stub>
  <v-footer-stub app="true" color="teal darken-2" height="32"><span class="footer-text text-xs-center white--text">Â© 2019</span></v-footer-stub>
</v-app-stub>
`;

exports[`App component renders error if query fails 1`] = `
<v-app-stub id="app">
  <v-toolbar-stub fixed="true" app="true" color="teal darken-2" dark="true" scrollthreshold="300"><img src="./assets/vue-graphql.png" alt="Logo" class="logo">
    <v-toolbar-title-stub class="hidden-xs-only">Vue Heroes</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-content-stub tag="main" class="teal lighten-3">
    <v-container-stub tag="div" fluid="" fill-height="" grid-list-md="" class="app-container white">
      <h2 class="test-error">
        Something went wrong. Please try again in a minute
      </h2>
    </v-container-stub>
  </v-content-stub>
  <v-footer-stub app="true" color="teal darken-2" height="32"><span class="footer-text text-xs-center white--text">Â© 2019</span></v-footer-stub>
</v-app-stub>
`;
